File: C/life.c
#include <stdio.h>
#include <stdlib.h>

#define DEFAULT 10

char** getInitialWorld(char* filename, int row, int column); 
void nextGeneration(char*** world, char*** newWorld, int row, int column);
void printWorld(int generationCount, char** map, int row, int column);
char** allocateWorld(int row, int column);
void freeWorld(char** world, int row);

int main(int argc, char *argv[]) {
	// Default values
	char* filename = "../input.txt";
	int row = DEFAULT;
	int column = DEFAULT;
	int generation = DEFAULT;

	// Check if input is vaild, cant take 6+ arguments
	if (argc > 1) {
		if (argc > 4) generation = atoi(argv[4]);
		if (argc > 3) filename = (char*)argv[3];
		if (argc > 2) column = atoi(argv[2]);
		row = atoi(argv[1]);
	}
	if (row == 0 || column == 0 || generation == 0 || argc > 5) {
		printf ("Error! invalid arguments\n");
		exit(1);
	}

	char** world = getInitialWorld(filename, row, column); 
	char** newWorld = allocateWorld(row, column);

	// Process the world
	for (int i = 0; i <= generation; ++i) {
		printWorld(i, world, row, column);
		nextGeneration(&world, &newWorld, row, column);
	}

	freeWorld(world, row);
	freeWorld(newWorld, row);
	return 0;
}

char** getInitialWorld(char* filename, int row, int column) {
	// Open the file and read
	FILE* file = fopen(filename, "r");
	if (file == NULL) {
		printf("Error! opening file\n");
		exit(1);
	}

	char **world = allocateWorld(row, column);

	// Read from file
	for (int i = 0; i < row; ++i) {
		char chr = getc(file);
		for (int j = 0; j < column; ++j) {
			if (chr != EOF && chr != '\n'){
				world[i][j] = chr;
				chr = getc(file);
			} else
				// When the line is too short, fill the row with dead cells
				world[i][j] = ' ';
		}
		// When a line is too long, overlook the extra ones
		while (chr != EOF && chr != '\n') chr = getc(file);
	}
	fclose(file);

	return world;
}

void printWorld(int generationCount, char** map, int row, int column) {
	printf("Generation %i:\n", generationCount);

	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < column; ++j) 
			(map[i][j] == ' ') ? printf("%c", '-') : printf("%c", map[i][j]);
		printf("\n");
	}

	printf("========================\n");
};

// Check if a cell is valid
int isValid(int r, int c, int row, int column) {
	return (r >= 0 && r < row && c >= 0 && c < column);
}

// Count neighbors
int countNeighbors(char** world, int i, int j, int row, int column) {
	int count = 0;
	count += (isValid(i - 1, j - 1, row, column) && world[i - 1][j - 1] == '*');
	count += (isValid(i - 1, j, row, column) && world[i - 1][j] == '*');
	count += (isValid(i - 1, j + 1, row, column) && world[i - 1][j + 1] == '*');
	count += (isValid(i, j - 1, row, column) && world[i][j - 1] == '*');
	count += (isValid(i, j + 1, row, column) && world[i][j + 1] == '*');
	count += (isValid(i + 1, j - 1, row, column) && world[i + 1][j - 1] == '*');
	count += (isValid(i + 1, j, row, column) && world[i + 1][j] == '*');
	count += (isValid(i + 1, j + 1, row, column) && world[i + 1][j + 1] == '*');
	return count;
}

void nextGeneration(char*** world, char*** newWorld, int row, int column) {
	// Decide alive or dead
	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < column; ++j) {
			int aliveCount = countNeighbors(*world, i, j, row, column);
			
			if (aliveCount < 2) 
				(*newWorld)[i][j] = ' ';
			else if (aliveCount > 3 && (*world)[i][j] == '*')
				(*newWorld)[i][j] = ' '; 
			else if (aliveCount == 3 && (*world)[i][j] == ' ') 
				(*newWorld)[i][j] = '*'; 
			else 
				(*newWorld)[i][j] = (*world)[i][j]; 
		}
	}
	
	// Copy over from newWorld
	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < column; ++j)
			(*world)[i][j] = (*newWorld)[i][j]; 
	}
}

char** allocateWorld(int row, int column) {
	char** world = (char **)malloc(row * sizeof(char*));

	for (int i = 0; i < row; i++) {
		world[i] = (char *)malloc(column * sizeof(char));
		if (!world[i]) {
			printf ("Error! allocate heap space\n");
			exit(1);
		}
	}
		
	if (!world) {
		printf ("Error! allocate heap space\n");
		exit(1);
	}

	return world;
}

void freeWorld(char** world, int row){
	for (int i = 0; i < row; i++)
		free(world[i]);
	free(world);
}

File: PHP/life.php
<?php
define("SIDE", 10);

function getInitialWorld() {
    $world = array();
    $myfile = fopen("../input.txt", "r") or die("Unable to open file!");

    for ($r = 0; $r < SIDE; $r++) {
        $row = array();
        $char = fgetc($myfile);
        for ($c = 0; $c < SIDE; $c++) {
            if (!feof($myfile) && ($char == ' ' || $char == '*')) {
                array_push($row, $char);
                $char = fgetc($myfile);
            } else {
                array_push($row, ' ');
            }
        }
        array_push($world, $row);
    }
    fclose($myfile);

    return $world;
}

function printWorld($generationCount, $world) {
    echo "Generation $generationCount:<br>";
    foreach ( $world as $row ) {
        foreach ( $row as $col ) {
            if ($col == ' ') $col = ' -';
            echo $col; 
        }
        echo "<br>";
    }
    echo "========================<br>";
}

function isValidCell($row, $col) {
	return ($row >= 0 && $row < SIDE && $col >= 0 && $col < SIDE);
}

function neighborNum($row, $col, $world) {
	$count = 0;
	for ($r = -1; $r <= 1; $r++) {
		for ($c = -1; $c <= 1; $c++) {
			if ($r == 0 && $c == 0) continue;
			if (isValidCell($row + $r, $col + $c) && $world[$row + $r][$col + $c] == '*')
                $count++;
		}
	}
	return $count;
}

function nextGeneration($world) {
    $newWorld = array();
    
    for ($r = 0; $r < SIDE; $r++) {
        $row = array();
        for ($c = 0; $c < SIDE; $c++) {
            $aliveCount = neighborNum($r, $c, $world);
            if ($aliveCount < 2)
                array_push($row, ' ');
			else if ($aliveCount > 3 && $world[$r][$c] == '*')
                array_push($row, ' ');
			else if ($aliveCount == 3 && $world[$r][$c] == ' ')
                array_push($row, '*');
            else 
                array_push($row, $world[$r][$c]);
        }
        array_push($newWorld, $row);
    }

    return $newWorld;
}

$world = getInitialWorld(); 
for ($i = 0; $i <= $_POST['generation']; $i++) {
    printWorld($i, $world);
    $world = nextGeneration($world);
}

?>